### PR TITLE
ft: Helm chart for backbeat lifecycle

### PR DESCRIPTION
**Pull request by tmacro**

### CODE REVIEW

**Comment by vrancurel** [comment]
@rahulreddy do we name the charts backbeat-lifecycle or lifecycle ?

**Comment by rahulreddy** [comment]
I think we should move away from calling the components backbeat-*. Just lifecycle,replication etc should be ok. This brings us to backbeat-api, should we call it zenko-api/workflow-api instead?

File: charts/backbeat-lifecycle/templates/conductor/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,68 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "backbeat.fullname" . }}-conductor-lifecycle
+  labels:
+    app: {{ template "backbeat.name" . }}
```
Comment (giacomoguiulfo) [inline charts/backbeat-lifecycle/templates/conductor/deployment.yaml]: Maybe the ```app``` label should be ```{{ template "backbeat.name" . }}-lifecycle```?

File: charts/backbeat-lifecycle/templates/conductor/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    chart: {{ template "backbeat.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+spec:
+  replicas: {{ .Values.conductor.replicaCount }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "backbeat.name" . }}
```
Comment (giacomoguiulfo) [inline charts/backbeat-lifecycle/templates/conductor/deployment.yaml]: Same here.

File: charts/backbeat-lifecycle/templates/consumer/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,68 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "backbeat.fullname" . }}-consumer-lifecycle
+  labels:
+    app: {{ template "backbeat.name" . }}
```
Comment (giacomoguiulfo) [inline charts/backbeat-lifecycle/templates/consumer/deployment.yaml]: Same here.

File: charts/backbeat-lifecycle/templates/consumer/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    chart: {{ template "backbeat.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+spec:
+  replicas: {{ .Values.consumer.replicaCount }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "backbeat.name" . }}
```
Comment (giacomoguiulfo) [inline charts/backbeat-lifecycle/templates/consumer/deployment.yaml]: Same here.

File: charts/backbeat-lifecycle/templates/conductor/deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        release: {{ .Release.Name }}
+    spec:
+      containers:
+        - name: {{ .Chart.Name }}
```
Comment (giacomoguiulfo) [inline charts/backbeat-lifecycle/templates/conductor/deployment.yaml]: It might be best if the container name matches the pod instead of just ```backbeat```.  Like this ```{{ template "backbeat.fullname" . }}-conductor-lifecycle```

File: charts/backbeat-lifecycle/templates/consumer/deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        release: {{ .Release.Name }}
+    spec:
+      containers:
+        - name: {{ .Chart.Name }}
```
Comment (giacomoguiulfo) [inline charts/backbeat-lifecycle/templates/consumer/deployment.yaml]: Same for this container name.

File: charts/backbeat-lifecycle/templates/producer/deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+spec:
+  replicas: {{ .Values.producer.replicaCount }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "backbeat.name" . }}
+        release: {{ .Release.Name }}
+    spec:
+      containers:
+        - name: {{ .Chart.Name }}
```
Comment (giacomoguiulfo) [inline charts/backbeat-lifecycle/templates/producer/deployment.yaml]: Same for this container name.

File: charts/backbeat-lifecycle/templates/producer/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,68 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "backbeat.fullname" . }}-producer-lifecycle
+  labels:
+    app: {{ template "backbeat.name" . }}
```
Comment (giacomoguiulfo) [inline charts/backbeat-lifecycle/templates/producer/deployment.yaml]: Same change here for the ```app``` label.

File: charts/backbeat-lifecycle/templates/producer/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    chart: {{ template "backbeat.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+spec:
+  replicas: {{ .Values.producer.replicaCount }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "backbeat.name" . }}
```
Comment (giacomoguiulfo) [inline charts/backbeat-lifecycle/templates/producer/deployment.yaml]: Same here.

File: charts/backbeat/templates/lifecycle/conductor_deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        - name: {{ template "backbeat.fullname" . }}-lifecycle-conductor
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          command: ["/usr/src/app/docker-entrypoint.sh"]
+          args: ["npm", "run", "lifecycle_conductor"]
+          env:
+            - name: ZOOKEEPER_AUTO_CREATE_NAMESPACE
+              value: "1"
+            - name: ZOOKEEPER_CONNECTION_STRING
+              value: "{{- printf "%s-zenko-zookeeper:2181" .Release.Name | trunc 63 | trimSuffix "-" -}}"
```
Comment (giacomoguiulfo) [inline charts/backbeat/templates/lifecycle/conductor_deployment.yaml]: Should be ```zenko-quorum``` instead of ```zenko-zookeeper```.

File: charts/backbeat/templates/lifecycle/conductor_deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+            - name: KAFKA_HOSTS
+              value: "{{- printf "%s-kafka-0.%s-kafka:9092" .Release.Name .Release.Name | trunc 63 | trimSuffix "-" -}}"
```
Comment (giacomoguiulfo) [inline charts/backbeat/templates/lifecycle/conductor_deployment.yaml]: ```zenko-quorum``` instead of ```kafka```.

File: charts/backbeat/templates/lifecycle/consumer_deployment.yaml [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        - name: {{ template "backbeat.fullname" . }}-lifecycle-consumer
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          command: ["/usr/src/app/docker-entrypoint.sh"]
+          args: ["npm", "run", "lifecycle_consumer"]
+          env:
+            - name: ZOOKEEPER_AUTO_CREATE_NAMESPACE
+              value: "1"
+            - name: ZOOKEEPER_CONNECTION_STRING
+              value: "{{- printf "%s-zenko-zookeeper:2181" .Release.Name | trunc 63 | trimSuffix "-" -}}"
```
Comment (giacomoguiulfo) [inline charts/backbeat/templates/lifecycle/consumer_deployment.yaml]: Same here.

File: charts/backbeat/templates/lifecycle/consumer_deployment.yaml [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+            - name: KAFKA_HOSTS
+              value: "{{- printf "%s-kafka-0.%s-kafka:9092" .Release.Name .Release.Name | trunc 63 | trimSuffix "-" -}}"
```
Comment (giacomoguiulfo) [inline charts/backbeat/templates/lifecycle/consumer_deployment.yaml]: Same here.

File: charts/backbeat/templates/lifecycle/producer_deployment.yaml [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          command: ["/usr/src/app/docker-entrypoint.sh"]
+          args: ["npm", "run", "lifecycle_producer"]
+          env:
+            - name: ZOOKEEPER_AUTO_CREATE_NAMESPACE
+              value: "1"
+            - name: ZOOKEEPER_CONNECTION_STRING
+              value: "{{- printf "%s-zenko-quorum:2181" .Release.Name | trunc 63 | trimSuffix "-" -}}"
+            - name: KAFKA_HOSTS
+              value: "{{- printf "%s-kafka-0.%s-kafka:9092" .Release.Name .Release.Name | trunc 63 | trimSuffix "-" -}}"
```
Comment (ssalaues) [inline charts/backbeat/templates/lifecycle/producer_deployment.yaml]: Should be able to use just the service name now as in "%s-zenko-queue:9092"

File: charts/backbeat/templates/lifecycle/conductor_deployment.yaml [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+spec:
+  replicas: {{ .Values.lifecycle.conductor.replicaCount }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "backbeat.name" . }}-lifecycle
+        release: {{ .Release.Name }}
+    spec:
+      containers:
+        - name: {{ template "backbeat.fullname" . }}-lifecycle-conductor
```
Comment (NicolasT) [inline charts/backbeat/templates/lifecycle/conductor_deployment.yaml]: No need to use long names for containers. Use a descriptive name instead.

File: charts/backbeat/templates/lifecycle/conductor_deployment.yaml [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+              value: "{{- printf "%s-kafka-0.%s-kafka:9092" .Release.Name .Release.Name | trunc 63 | trimSuffix "-" -}}"
+            - name: MONGODB_HOSTS
+              value: "{{ .Release.Name }}-mongodb-replicaset-0.{{ .Release.Name }}-mongodb-replicaset:27017,{{ .Release.Name }}-mongodb-replicaset-1.{{ .Release.Name }}-mongodb-replicaset:27017,{{ .Release.Name }}-mongodb-replicaset-2.{{ .Release.Name }}-mongodb-replicaset:27017"
+            - name: AUTH_TYPE
+              value: service
+            - name: AUTH_ACCOUNT
+              value: service-lifecycle
+            - name: EXTENSIONS_LIFECYCLE_ZOOKEEPER_PATH
+              value: "{{ .Values.lifecycle.zookeeper.path }}"
+            - name: EXTENSIONS_LIFECYCLE_BACKLOG_METRICS_ZKPATH
```
Comment (NicolasT) [inline charts/backbeat/templates/lifecycle/conductor_deployment.yaml]: What's the relation between metrics and Zookeeper?

File: charts/backbeat/templates/lifecycle/conductor_deployment.yaml [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+              value: "{{ .Values.lifecycle.zookeeper.backlog_metrics.path }}"
+            - name: EXTENSIONS_LIFECYCLE_BACKLOG_METRICS_INTERVALS
+              value: "{{ .Values.lifecycle.zookeeper.backlog_metrics.interval }}"
+            - name: EXTENSIONS_LIFECYCLE_BUCKET_TASK_TOPIC
+              value: "{{ .Values.lifecycle.kafka.bucket_task_topic }}"
+            - name: EXTENSIONS_LIFECYCLE_OBJECT_TASK_TOPIC
+              value: "{{ .Values.lifecycle.kafka.object_task_topic }}"
+            - name: EXTENSIONS_LIFECYCLE_CONDUCTOR_CRONRULE
```
Comment (NicolasT) [inline charts/backbeat/templates/lifecycle/conductor_deployment.yaml]: If we are doing cron'y things in our solution, wouldn't it make sense to use Kubernetes' `CronJob` objects?

File: charts/backbeat/values.yaml [hunk 18]
```diff
@@ -31,6 +31,49 @@ api:
...
+
+  producer:
+    replicaCount: 1
+    group: backbeat-lifecycle-producer-group
+    resources: {}
+    nodeSelector: {}
+    tolerations: []
+    affinity: {}
+
+  zookeeper:
```
Comment (NicolasT) [inline charts/backbeat/values.yaml]: On one hand, some want to use 'functional' names for objects etc. instead of service names. On the other hand, we use service names in configuration. This is... confusing. And also, hard to change later.

File: charts/backbeat/templates/lifecycle/conductor_deployment.yaml (hunk 17 above)
Comment (jonathan-gramain) [reply charts/backbeat/templates/lifecycle/conductor_deployment.yaml]: I have no particular knowledge of cron jobs in kub but why not since it might be easier to control and visualize, though I think it's better to do this in a next step (more code changes to conductor process than just deployment settings).

File: charts/backbeat/values.yaml [hunk 19]
```diff
@@ -31,6 +31,49 @@ api:
...
+    path: "/lifecycle"
+    backlog_metrics:
+      path: "/lifecycle/run/backlog-metrics"
+      interval: 60
+
+  kafka:
+    bucket_task_topic: backbeat-lifecycle-bucket-tasks
```
[the remark this answers is not in the document]
Comment (jonathan-gramain) [reply charts/backbeat/values.yaml]: Now that I think about it, I think it's better not to have the topic names in `values.yaml` as those are only for internal use and should not be modified by the admin in normal circumstances (so having them declared with their actual topic name in each chart would be better).

File: charts/backbeat/templates/lifecycle/conductor_deployment.yaml (hunk 16 above)
Comment (tmacro) [reply charts/backbeat/templates/lifecycle/conductor_deployment.yaml]: The name is taken from the backbeat config file 
```
"extensions": {
    "lifecycle": {
        ...
        "backlogMetrics": {
            "zkPath": "/lifecycle/run/backlog-metrics",
            ....
        },
    },
}
```

File: charts/backbeat/templates/lifecycle/producer_deployment.yaml [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          command: ["/usr/src/app/docker-entrypoint.sh"]
+          args: ["npm", "run", "lifecycle_producer"]
+          env:
+            - name: ZOOKEEPER_AUTO_CREATE_NAMESPACE
+              value: "1"
+            - name: ZOOKEEPER_CONNECTION_STRING
+              value: "{{- printf "%s-zenko-quorum:2181" .Release.Name | trunc 63 | trimSuffix "-" -}}"
+            - name: KAFKA_HOSTS
+              value: "{{- printf "%s-zenko-0.%s-queue:9092" .Release.Name .Release.Name | trunc 63 | trimSuffix "-" -}}"
```
Comment (giacomoguiulfo) [inline charts/backbeat/templates/lifecycle/producer_deployment.yaml]: Small typo. It should look like this```...printf "%s-zenko-queue:9092" .Release.Name...```

File: charts/backbeat/templates/lifecycle/conductor_deployment.yaml [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+            - name: AUTH_ACCOUNT
+              value: service-lifecycle
+            - name: EXTENSIONS_LIFECYCLE_ZOOKEEPER_PATH
+              value: "{{ .Values.lifecycle.zookeeper.path }}"
+            - name: EXTENSIONS_LIFECYCLE_BACKLOG_METRICS_ZKPATH
+              value: "{{ .Values.lifecycle.zookeeper.backlog_metrics.path }}"
+            - name: EXTENSIONS_LIFECYCLE_BACKLOG_METRICS_INTERVALS
+              value: "{{ .Values.lifecycle.zookeeper.backlog_metrics.interval }}"
+            - name: EXTENSIONS_LIFECYCLE_BUCKET_TASK_TOPIC
+              value: bucket_tasks_topic
```
Comment (jonathan-gramain) [inline charts/backbeat/templates/lifecycle/conductor_deployment.yaml]: The actual topic name we use is `backbeat-lifecycle-bucket-tasks`, should be set here and for other lifecycle components.

File: charts/backbeat/templates/lifecycle/conductor_deployment.yaml [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+            - name: EXTENSIONS_LIFECYCLE_OBJECT_TASK_TOPIC
+              value: object_tasks_topic
```
Comment (jonathan-gramain) [inline charts/backbeat/templates/lifecycle/conductor_deployment.yaml]: Topic name is `backbeat-lifecycle-object-tasks`

**Comment by rahulreddy** [comment]
Where do we stand on this PR?

**Comment by giacomoguiulfo** [comment]
@rahulreddy, as far as I know, there needs to be a change in the lifecycle code before this chart works. Something related to service accounts if I'm not wrong.

File: charts/backbeat/templates/lifecycle/producer_deployment.yaml [hunk 23]
```diff
@@ -0,0 +1,74 @@
+apiVersion: extensions/v1beta1
```
Comment (NicolasT) [inline charts/backbeat/templates/lifecycle/producer_deployment.yaml]: This looks like 3x almost-the-same file. Consider abstracting out the template, parametrize it and run it 3 times?

Comment (ssalaues) [reply charts/backbeat/templates/lifecycle/producer_deployment.yaml]: Definitely agree with Nicolas but lets leave that for another PR. I wanted to separate the burden of the refactor and the addition of this feature. We have a ticket ZENKO-388 specifically to do exactly what Nicolas is speaking of. 😄

File: charts/backbeat/templates/replication/consumer_deployment.yaml [hunk 24]
```diff
@@ -30,6 +30,8 @@ spec:
               value: "{{- printf "%s-zenko-quorum:2181" .Release.Name | trunc 63 | trimSuffix "-" -}}"
             - name: KAFKA_HOSTS
               value: "{{- printf "%s-zenko-queue:9092" .Release.Name | trunc 63 | trimSuffix "-" -}}"
+            - name: LOG_LEVEL
```
Comment (vrancurel) [inline charts/backbeat/templates/replication/consumer_deployment.yaml]: this shall be in another PR. we want to avoid silent fixes

File: charts/backbeat/templates/replication/producer_deployment.yaml [hunk 25]
```diff
@@ -35,6 +35,8 @@ spec:
               value: "{{- printf "%s-%s" .Release.Name "redis-ha-master-svc" | trunc 63 | trimSuffix "-" -}}"
             - name: REDIS_PORT
               value: "6379"
+            - name: LOG_LEVEL
```
Comment (vrancurel) [inline charts/backbeat/templates/replication/producer_deployment.yaml]: same, silent fix => another PR

File: charts/backbeat/templates/replication/status_deployment.yaml [hunk 26]
```diff
@@ -44,6 +44,8 @@ spec:
               value: "{{- printf "%s-%s" .Release.Name "redis" | trunc 63 | trimSuffix "-" -}}"
             - name: REDIS_PORT
               value: "6379"
+            - name: LOG_LEVEL
+              value: {{ .Values.logging.level }}
```
Comment (vrancurel) [inline charts/backbeat/templates/replication/status_deployment.yaml]: silent fix

**Comment by ssalaues** [comment]
We are waiting on https://github.com/scality/backbeat/pull/299 before merging.

**Comment by vrancurel** [comment]
tested manually, bumped into ZENKO-481

**Comment by ssalaues** [comment]
It looks like bumping to the new cloudserver causes the the CI tests to fail. Is it a regression possibly?
